core/lib/node_framework/src/resource/mod.rs

```diff
@@ -23,40 +48,42 @@ mod unique;
 ///     fn do_something(&self);
 /// }
 ///
-/// /// Resource wrapper.
-/// #[derive(Clone)]
-/// struct MyResource(Arc<dyn MyInterface>);
-///
-/// impl Resource for MyResource {
+/// impl Resource<Shared> for dyn MyInterface {
 ///     fn name() -> String {
 ///         // It is a helpful practice to follow a structured naming pattern for resource names.
 ///         // For example, you can use a certain prefix for all resources related to a some component, e.g. `api`.
 ///         "common/my_resource".to_string()
 ///     }
 /// }
+///
+/// // The resource can now be injected / requested as `Arc<dyn MyInterface>`
 /// ```
-pub trait Resource: 'static + Send + Sync + std::any::Any {
-    /// Invoked after the wiring phase of the service is done.
-    /// Can be used to perform additional resource preparation, knowing that the resource
-    /// is guaranteed to be requested by all the tasks that need it.
-    fn on_resource_wired(&mut self) {}
-
+pub trait Resource<Kind: ResourceKind = Plain>: 'static + Send + Sync {
```

Comment:
Should we add `#[diagnostic::on_unimplemented]` here to mention that if trait is not implemented, but you expect it to, one must check the kind of resource (Plain/Shared/Boxed) and mention differences?